src/lib/render.ts

```diff
@@ -1,166 +1,179 @@
+	}
+
+	const source_index = 3 * (lowres_picker_size * lowres_picker_size - 1)
+	const target_index = 4 * (picker_size * picker_size - 1)
+
+	const r = lowres_data[source_index + 0]
+	const g = lowres_data[source_index + 1]
+	const b = lowres_data[source_index + 2]
+
+	data[target_index + 0] = r
+	data[target_index + 1] = g
+	data[target_index + 2] = b
 }
 
-export function render_main_image(hue: number, image: ImageData, transform: (rgb: number[]) => number[]) {
-  const data = image.data
-  const lowres_data = new Float32Array(lowres_picker_size * lowres_picker_size * 3)
-
-  for (let i = 0; i < lowres_picker_size; i++) {
-    for (let j = 0; j < lowres_picker_size; j++) {
-      const rgb = transform(okhsv_to_srgb(hue, 2 * j * picker_size_inv, 1 - 2 * i * picker_size_inv))
-
-      const index = 3 * (i * lowres_picker_size + j)
-      lowres_data[index + 0] = rgb[0]
-      lowres_data[index + 1] = rgb[1]
-      lowres_data[index + 2] = rgb[2]
-    }
-  }
-
-  for (let i = 0; i < picker_size; i++) {
-    for (let j = 0; j < picker_size; j++) {
-      const index = 4 * (i * picker_size + j)
-      data[index + 3] = 255
-    }
-  }
-
-  upscale(lowres_data, data)
+export function render_main_image(
+	hue: number,
+	image: ImageData,
+	transform: (rgb: number[]) => number[]
+) {
+	const data = image.data
+	const lowres_data = new Float32Array(
+		lowres_picker_size * lowres_picker_size * 3
+	)
+
+	for (let i = 0; i < lowres_picker_size; i++) {
+		for (let j = 0; j < lowres_picker_size; j++) {
+			const rgb = toRgb({
+				mode: 'okhsv',
+				h: hue,
+				s: 2 * j * picker_size_inv,
+				v: 1 - 2 * i * picker_size_inv,
+			})
+
+			const index = 3 * (i * lowres_picker_size + j)
+			lowres_data[index + 0] = rgb.r * 255
+			lowres_data[index + 1] = rgb.g * 255
+			lowres_data[index + 2] = rgb.b * 255
+		}
+	}
+
+	for (let i = 0; i < picker_size; i++) {
+		for (let j = 0; j < picker_size; j++) {
+			const index = 4 * (i * picker_size + j)
+			data[index + 3] = 255
+		}
+	}
+
+	upscale(lowres_data, data)
 }
 
 export function render_slider_image() {
-  const data = new Uint8ClampedArray(picker_size * slider_width * 4)
-
-  for (let i = 0; i < picker_size; i++) {
-    const a_ = Math.cos(2 * Math.PI * i * picker_size_inv)
-    const b_ = Math.sin(2 * Math.PI * i * picker_size_inv)
-
-    const rgb = hsluv_to_rgb(i * picker_size_inv, 0.9, 0.65 + 0.20 * b_ - 0.09 * a_)
-
-    for (let j = 0; j < slider_width; j++) {
-      const index = 4 * (i * slider_width + j)
-      data[index + 0] = rgb[0]
-      data[index + 1] = rgb[1]
-      data[index + 2] = rgb[2]
-      data[index + 3] = 255
-    }
-  }
-
-  return new ImageData(data, slider_width)
-}
\ No newline at end of file
+	const data = new Uint8ClampedArray(picker_size * slider_width * 4)
+
+	for (let i = 0; i < picker_size; i++) {
+		const a_ = Math.cos(2 * Math.PI * i * picker_size_inv)
+		const b_ = Math.sin(2 * Math.PI * i * picker_size_inv)
+
+		const rgb = toRgb({
```

Comment:
… and right here (2). It was pretty much a piece of cake to make the switch because Culori supports the same colorspace.

But it did result in a little bit of code deletion / simplifcation, which is nice